pkg/cni/plugin_test.go

```diff
@@ -0,0 +1,195 @@
+package cni_test
+
+import (
+	"fmt"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/plugins/pkg/ns"
+	"github.com/containernetworking/plugins/pkg/testutils"
+	"github.com/kubevirt/macvtap-cni/pkg/cni"
+	"github.com/kubevirt/macvtap-cni/pkg/util"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/vishvananda/netlink"
+)
+
+const LOWER_DEVICE = "eth0"
+
+var _ = Describe("Macvtap device plugin", func() {
+	var originalNS ns.NetNS
+	var targetNs ns.NetNS
+	var lowerDevice netlink.Link
+	var macvtapInterface netlink.Link
+	var deviceID string
+	var stdInArgs string
+
+	BeforeEach(func() {
+		var err error
+		originalNS, err = testutils.NewNS()
+		Expect(err).NotTo(HaveOccurred())
+		targetNs, err = testutils.NewNS()
+		Expect(err).NotTo(HaveOccurred())
+
+		err = originalNS.Do(func(ns.NetNS) error {
+			defer GinkgoRecover()
+
+			// create lower device
+			err = netlink.LinkAdd(&netlink.Dummy{
+				LinkAttrs: netlink.LinkAttrs{
+					Name: LOWER_DEVICE,
+				},
+			})
+			Expect(err).NotTo(HaveOccurred())
+			lowerDevice, err = netlink.LinkByName(LOWER_DEVICE)
+			Expect(err).NotTo(HaveOccurred())
+
+			// create macvtap on top of lower device
+			macvtapIfaceName := "macvtap0"
+			_, err = util.CreateMacvtap(macvtapIfaceName, LOWER_DEVICE, "bridge")
+			Expect(err).NotTo(HaveOccurred())
+			macvtapInterface, err = netlink.LinkByName(macvtapIfaceName)
+			Expect(err).NotTo(HaveOccurred())
+			deviceID = macvtapIfaceName
+
+			stdInArgs = fmt.Sprintf(`{
+				"cniVersion": "0.3.1",
+				"name": "mynet",
+				"type": "macvtap",
+				"deviceID": "%s"
+			}`, macvtapIfaceName)
+			return nil
+		})
+
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		// cleanup the original namespace
+		originalNS.Do(func(ns.NetNS) error {
+			return netlink.LinkDel(lowerDevice)
+		})
+		Expect(originalNS.Close()).To(Succeed())
+		Expect(testutils.UnmountNS(originalNS)).To(Succeed())
+
+		// cleanup the target namespace
+		targetNs.Do(func(ns.NetNS) error {
+			return netlink.LinkDel(macvtapInterface)
+		})
+		Expect(targetNs.Close()).To(Succeed())
+		Expect(testutils.UnmountNS(targetNs)).To(Succeed())
+	})
+
+	It("imports a macvtap interface into the target netns", func() {
+		args := &skel.CmdArgs{
```

Comment:
This should have a inner `Context("after executing the CNI")`. The execution should be in BeforeEach here. Alternatively you don't need the second Context, but then name the `It` "should successfully import ... while executing Add" or something similar.

Just try to follow Contexts, BeforeEach and Its, read it as a sentence and make sure that contents of each really do what the part of the sentence describes. That should make it easier to 1) follow output of the test run 2) read the code.

Reply:
Split it into 3 different contexts.

Done.